File: code/go/01_array/sort_square/sort_square.go
package sort_square

// 有序数组的平方 https://leetcode.cn/problems/squares-of-a-sorted-array/

// 双指针法
// 时间复杂度：O(1)
func SortSquare1(nums []int) []int {
	len := len(nums)
	l, r, i := 0, len-1, len-1
	res := make([]int, len)
	for l <= r {
		ln, rn := nums[l]*nums[l], nums[r]*nums[r]
		if ln < rn {
			res[i] = rn
			r--
		} else {
			res[i] = ln
			l++
		}
		i--
	}
	return res
}

File: code/php/01_array/sort_square/sort_square.php
<?php

/**
 * 有序数组的平方 https://leetcode.cn/problems/squares-of-a-sorted-array/
 */
class SortSquare
{
    /**
     * 有序数组的平方
     *
     * @param array $nums
     * @return array
     */
    public function sortSquare(array $nums): array
    {
        if (empty($nums)) {
            return $nums;
        }

        $res = [];
        for ($i = 0; $i < count($nums); $i++) {
            $res[$i] = 0;
        }

        $index = count($nums) - 1;

        for ($i = 0, $j = count($nums) - 1; $i <= $j;) {
            if ($nums[$i] ** 2 > $nums[$j] ** 2) {
                $res[$index] = $nums[$i] ** 2;
                $i++;
            } else {
                $res[$index] = $nums[$j] ** 2;
                $j--;
            }

            $index--;
        }

        return $res;
    }
}


$sortSquare = new SortSquare();
$nums = [-4, -1, 0, 3, 10];
print_r($sortSquare->sortSquare($nums));

$nums =  [-7, -3, 2, 3, 11];
print_r($sortSquare->sortSquare($nums));